### PR TITLE
Alerting: Coremodel POC

### PR DESCRIPTION
**Pull request by joeblubaugh**
This is a *rough draft* for what a coremodel / schema for an alerting rule would look like, based on the type used by the Grafana Provisioning API, the ProvisionedAlertRule.

I cribbed off of dashboards a bit and I think:
1. We need a shared time library for cue, including both ISO-XYZ timestamps and relative durations. Maybe we have one and I missed it.
2. We need a shared definition of a key-value thing.
3. I would really like to figure out how to have bound constraints like "matches a regexp." It would be insanely useful.

There's a gap in cue: there isn't a straightforward way to declare something that generates a `map[string]string`, which alerting uses for its Label and Annotation types. Best I could hack together is a list of key-value pairs, but all users need to agree on deduplication rules.

@grafana/alerting-squad-frontend check out the generated typescript. Does it seem reasonable?

@sdboyer I haven't been able to get a clear sense from the existing code yet. What would validating arbitrary JSON against this schema look like in an API call?

### CODE REVIEW

**Comment by sdboyer** [comment]
woohoo, amazing! quick first response:

> There's a gap in cue: there isn't a straightforward way to declare something that generates a map[string]string, which alerting uses for its Label and Annotation types.

(i'm assuming you tried a [template](https://cuelang.org/docs/tutorials/tour/types/templates/))

Hmm, i wonder if this is a deficiency in the openapi generator. I know that, at least, [Thema can validate the assignability of maps to schemas](https://github.com/grafana/thema/blob/main/assignable_test.go#L119-L127).

> We need a shared definition of a key-value thing.

say more?

> I would really like to figure out how to have bound constraints like "matches a regexp." It would be insanely useful.

Ask and ye shall receive https://cuelang.org/docs/tutorials/tour/expressions/regexp/

The only drawback to these is that calculating subsumption of two regexps is infeasible, which means that any change to a regexp will ultimately be consider a backwards-incompatible change by Thema's invariant checker.

**Comment by joeblubaugh** [comment]
> Ask and ye shall receive https://cuelang.org/docs/tutorials/tour/expressions/regexp/

when I added one of these I didn’t see any change to generated code. The cue runtime would do the regex validation, though?

**Comment by sdboyer** [comment]
> when I added one of these I didn’t see any change to generated code. The cue runtime would do the regex validation, though?

yep, that's the only way to do it - same as with any bounds constraints that aren't expressible in Go's type system.

**Comment by joeblubaugh** [comment]
Oh, I never would have guessed that map syntax.

Sadly
```cue
labels: [string] : string
```
generates

```go
Labels: map[string]interface{}
```

Is that something we can help fix?

File: packages/grafana-schema/src/index.gen.ts [hunk 1]
```diff
@@ -4,6 +4,34 @@
...
+// The following exported declarations correspond to types in the alertrule@0.0 schema with
+// attribute @grafana(TSVeneer="type"). (lineage declared in file: pkg/coremodel/alertrule/coremodel.cue)
+//
+// The handwritten file for these type and default veneers is expected to be at
+// packages/grafana-schema/src/veneer/alertrule.types.ts.
+// This re-export declaration enforces that the handwritten veneer file exists,
+// and exports all the symbols in the list.
+//
+// TODO generate code such that tsc enforces type compatibility between raw and veneer decls
+export type { Alertrule } from './veneer/alertrule.types';
```
Comment (sdboyer) [inline packages/grafana-schema/src/index.gen.ts:22]: It's cool to see the veneer types in use again already! :)

looks like you haven't made `packages/grafana-schema/src/veneer/alertrule.types.ts` - you'll need to hand-create it and re-export the `Alertrule` symbol with whatever type additions you want to make

(this is definitely one of those things that will be in our tutorial)

**Comment by sdboyer** [comment]
> Oh, I never would have guessed that map syntax.

For sure - that's because of the different path by which CUE arrives at this destination. From CUE's perspective, these are just structs/a set of key/value pairs, which is CUE's foundational construct. So it's not a map any more than _everything_ in CUE is a map - what `field: [string]: string` is really saying is, "constrain that `field`'s value is a) a struct, b) with `string` fields, and c) `string` values".

> Is that something we can help fix?

i'm working up some more tests in thema for things that will hit this as part of grafana/thema#72, which should at least make it clear what current capabilities are. But how easy it is to fix/where to fix it will depend on exactly where the breakdown is happening - it's either in the CUE->OpenAPI transition, or OpenAPI->Go.

If it's the former, it's still not necessarily trivial, but we can put some mutators in [here](https://github.com/grafana/thema/blob/main/encoding/openapi/oapi.go) (analogous to what thema does to [make json schema out of openapi](https://github.com/grafana/thema/blob/main/encoding/jsonschema/jschema.go#L40)) that fix it up, while we submit a fix upstream to CUE itself.

If it's the latter, that may be harder - it'd be the [kin-openapi generator](github.com/getkin/kin-openapi/openapi3) that we depend on. Not sure, haven't tried submitting anything to them yet.

There's also a shorter-term hack available - we could introduce another CUE attribute in which the type of the output can be literally specified: `@grafana(forceGoType="map[string]string")`. 

Aside: I've previously argued against forcing type output in the TS context. The reason i'm comfortable with it here is because in Go, alignment between the generated type and the schema is also enforced at runtime by Thema's [AssignableTo()](https://pkg.go.dev/github.com/grafana/thema#AssignableTo) check. That's as good as a static type check, at least from the perspective of preventing bugs/abuse in having a dumb passthrough for the type.

**Comment by joeblubaugh** [comment]
> There's also a shorter-term hack available - we could introduce another CUE attribute in which the type of the output can be literally specified: @grafana(forceGoType="map[string]string").

I'd really like to avoid this, especially when we advocate thema outside of the company. These escape hatches make it possible to defer core improvements indefinitely when a project runs up against schedule pressure. If the problem is in kin-openapi then maybe that's the pragmatic solution but I hope not.

**Comment by sdboyer** [comment]
> I'd really like to avoid this, especially when we advocate thema outside of the company. These escape hatches make it possible to defer core improvements indefinitely when a project runs up against schedule pressure.

Couldn't agree more.

Fortunately, it looks like it's probably the former case - the bug is in [CUE stdlib](https://pkg.go.dev/cuelang.org/go/encoding/openapi). Given a simple input case to replicate:

```
$ cat > brokemap.cue <<EOF
package brokemap

import "github.com/grafana/thema"

thema.#Lineage
name: "brokemap"
seqs: [
	{
		schemas: [
			{
				stringstring: [string]: string
				stringint32: [string]: int32
			},
		]
	},
]
EOF
```

We can run some of the new gen hotness (not merged yet, grafana/thema#72) to confirm the broken Go type output:
```
$ thema lineage gen gotypes -l brokemap.cue
package brokemap

// Brokemap defines model for brokemap.
type Brokemap struct {
	Stringint32  map[string]interface{} `json:"stringint32"`
	Stringstring map[string]interface{} `json:"stringstring"`
}
```

Then more new gen hotness to see that the intermediate output from cue stdlib that's sent along to openapi is, itself, broken:

```
$ thema lineage gen openapi -l brokemap.cue
{
  "openapi": "3.0.0",
  "info": {
    "title": "brokemap",
    "version": "0.0"
  },
  "paths": {},
  "components": {
    "schemas": {
      "brokemap": {
        "type": "object",
        "required": [
          "stringstring",
          "stringint32"
        ],
        "properties": {
          "stringstring": {
            "type": "object"
          },
          "stringint32": {
            "type": "object"
          }
        }
      }
    }
  }
}
```

So, seems we're on the less-hard path. i know roughly how we'd approach fixing this by hacking the output from CUE stdlib in thema's openapi encoding package. Should be pretty straightforward if we restrict ourselves to instances where the map value has a primitive types.

**Comment by joeblubaugh** [comment]
@sdboyer do you want a ticket on the Thema repo?

**Comment by sdboyer** [comment]
yes please, I'd greatly appreciate that

**Comment by sdboyer** [comment]
Ugh, the problems extend to deepmap/oapi-codegen as well (that's the actual openapi code generator, we use kin-openapi for validation, not codegen - my mistake)

When i manually modify the openapi schema to be:
```json
{
  "openapi": "3.0.0",
  "info": {
    "title": "brokemap",
    "version": "0.0"
  },
  "paths": {},
  "components": {
    "schemas": {
      "brokemap": {
        "type": "object",
        "required": [
          "stringstring",
          "stringint32",
          "stringobj"
        ],
        "properties": {
          "stringstring": {
            "type": "object",
	    "additionalProperties": {
		    "type": "string"
	    }
          },
          "stringint32": {
            "type": "object"
          },
        }
      }
    }
  }
}
```
which, according to [this guide](https://swagger.io/docs/specification/data-models/dictionaries/), is the proper way to model a simple map, the code generator tries to get all fancy:
```go
// Package brokemap provides primitives to interact with the openapi HTTP API.
//
// Code generated by github.com/deepmap/oapi-codegen version v1.10.1 DO NOT EDIT.
package brokemap

import (
	"encoding/json"
	"fmt"
)

// Brokemap defines model for brokemap.
type Brokemap struct {
	Stringint32  map[string]interface{} `json:"stringint32"`
	Stringstring Brokemap_Stringstring  `json:"stringstring"`
}

// Brokemap_Stringstring defines model for Brokemap.Stringstring.
type Brokemap_Stringstring struct {
	AdditionalProperties map[string]string `json:"-"`
}

// Getter for additional properties for Brokemap_Stringstring. Returns the specified
// element and whether it was found
func (a Brokemap_Stringstring) Get(fieldName string) (value string, found bool) {
	if a.AdditionalProperties != nil {
		value, found = a.AdditionalProperties[fieldName]
	}
	return
}

// Setter for additional properties for Brokemap_Stringstring
func (a *Brokemap_Stringstring) Set(fieldName string, value string) {
	if a.AdditionalProperties == nil {
		a.AdditionalProperties = make(map[string]string)
	}
	a.AdditionalProperties[fieldName] = value
}

// Override default JSON handling for Brokemap_Stringstring to handle AdditionalProperties
func (a *Brokemap_Stringstring) UnmarshalJSON(b []byte) error {
	object := make(map[string]json.RawMessage)
	err := json.Unmarshal(b, &object)
	if err != nil {
		return err
	}

	if len(object) != 0 {
		a.AdditionalProperties = make(map[string]string)
		for fieldName, fieldBuf := range object {
			var fieldVal string
			err := json.Unmarshal(fieldBuf, &fieldVal)
			if err != nil {
				return fmt.Errorf("error unmarshaling field %s: %w", fieldName, err)
			}
			a.AdditionalProperties[fieldName] = fieldVal
		}
	}
	return nil
}

// Override default JSON handling for Brokemap_Stringstring to handle AdditionalProperties
func (a Brokemap_Stringstring) MarshalJSON() ([]byte, error) {
	var err error
	object := make(map[string]json.RawMessage)

	for fieldName, field := range a.AdditionalProperties {
		object[fieldName], err = json.Marshal(field)
		if err != nil {
			return nil, fmt.Errorf("error marshaling '%s': %w", fieldName, err)
		}
	}
	return json.Marshal(object)
}
```

There's no combination of options i can find with their generator to get it to stop doing all this. I'm assuming it does this because they're focused on providing the automated unmarshaler experience and the `Brokemap_Stringstring` standalone type is how they can reliably fit their generated logic into Go's native JSON unmarshaling interface system.

But that's not helpful for us - CUE is already doing our unmarshaling. It gives us the extra `Brokemap_Stringstring` type we don't want, with this explicit `AdditionalProperties` field that will fail Thema's assignability checker.

We _can_ solve this with AST manipulation of the deepmap output, though...it might get sticky. Targeting the generated funcs for removal with AST manipulation is straightforward - we simply delete all func-type AST nodes. Determining when to basically inline a type is harder, though i suspect we'll have a pretty strong clue by simply looking for fields with a `json:"-"` struct tag and/or the field name `AdditionalProperties`.

But it'd definitely be preferable if we could get them to accept a patch that introduces an option to just skip all of this indirection and unmarshaler generating logic. Without having looked at their code, that naively seems like it should be the thing to do - they're already getting the `map[string]string` part right (which is probably the bulk of the work). It "just" needs to be inlined.

File: pkg/coremodel/alertrule/coremodel.cue [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+				condition: string
+
+				queries?: [...#AlertQuery]
+
+				// TODO: timestamps
+				// updated
+
+				noDataState: #NoDataState
+
+				execErrState: #ExecutionErrState
```
Comment (sdboyer) [inline pkg/coremodel/alertrule/coremodel.cue:35]: Would i be correct in guessing that these are both things that, ideally, would be stored as metadata that only the system gets to write? That is, it'd be at best meaningless for someone trying to provision an alertrule to include values for these. 

cc @ryantxu

Comment (joeblubaugh) [reply pkg/coremodel/alertrule/coremodel.cue:35]: No, these are part of provisioning - you get to choose what the rule days if the final evaluation returns "ExecutionError" or "NoData". "Should I alert? Or should I say it's normal?"

**Comment by sdboyer** [comment]
fyi, we're finally chasing the map value gen issue down. grafana/schematization-and-as-code-project#25 for tracking

**Pull request closed without merging**